test/test_pairs_classifiers.py

```diff
@@ -85,3 +127,194 @@ def test_set_default_threshold_toy_example():
+  best_f1_score = f1_score(y, pairs_learner.predict(pairs))
+  for threshold in - pairs_learner.decision_function(pairs):
+    pairs_learner.set_threshold(threshold)
+    assert f1_score(y, pairs_learner.predict(pairs)) <= best_f1_score
+
+
+def true_pos_true_neg_rates(y_true, y_pred):
+  """A function that returns the true positive rates and the true negatives
+  rate. For testing purposes (optimized for readability not performance)."""
+  assert y_pred.shape[0] == y_true.shape[0]
+  tp = np.sum((y_pred == 1) * (y_true == 1))
+  tn = np.sum((y_pred == -1) * (y_true == -1))
+  fn = np.sum((y_pred == -1) * (y_true == 1))
+  fp = np.sum((y_pred == 1) * (y_true == -1))
+  tpr = tp / (tp + fn)
+  tnr = tn / (tn + fp)
+  tpr = tpr if not np.isnan(tpr) else 0.
+  tnr = tnr if not np.isnan(tnr) else 0.
+  return tpr, tnr
+
+
+def tpr_threshold(y_true, y_pred, tnr_threshold=0.):
+  """A function that returns the true positive rate if the true negative
+  rate is higher or equal than `threshold`, and -1 otherwise. For testing
+  purposes"""
+  tpr, tnr = true_pos_true_neg_rates(y_true, y_pred)
+  if tnr < tnr_threshold:
+    return -1
+  else:
+    return tpr
+
+
+def tnr_threshold(y_true, y_pred, tpr_threshold=0.):
+  """A function that returns the true negative rate if the true positive
+  rate is higher or equal than `threshold`, and -1 otherwise. For testing
+  purposes"""
+  tpr, tnr = true_pos_true_neg_rates(y_true, y_pred)
+  if tpr < tpr_threshold:
+    return -1
+  else:
+    return tnr
+
+
+@pytest.mark.parametrize('kwargs, scoring',
+                         [({'strategy': 'accuracy'}, accuracy_score),
+                          *[({'strategy': 'f_beta', 'beta': b},
+                             partial(fbeta_score, beta=b))
+                             for b in [0.1, 0.5, 1.]],
+                          ({'strategy': 'f_beta', 'beta': 0}, precision_score),
+                          *[({'strategy': 'max_tpr', 'threshold': t},
+                             partial(tpr_threshold, tnr_threshold=t))
+                             for t in [0., 0.1, 0.5, 0.8, 1.]],
+                          *[({'strategy': 'max_tnr', 'threshold': t},
+                             partial(tnr_threshold, tpr_threshold=t))
+                             for t in [0., 0.1, 0.5, 0.8, 1.]],
+                          ])
+def test_found_score_is_best_score(kwargs, scoring):
+  # test that when we use calibrate threshold, it will indeed be the
+  # threshold that have the best score
+  rng = np.random.RandomState(42)
+  n_samples = 50
+  pairs, y = rng.randn(n_samples, 2, 5), rng.choice([-1, 1], size=n_samples)
+  pairs_learner = IdentityPairsClassifier()
+  pairs_learner.fit(pairs, y)
+  pairs_learner.calibrate_threshold(pairs, y, **kwargs)
+  best_score = scoring(y, pairs_learner.predict(pairs))
+  scores = []
+  i = 0
+  predicted_scores = pairs_learner.decision_function(pairs)
+  predicted_scores = np.hstack([[np.min(predicted_scores) - 1],
```

Comment:
We also test that it works for thresholds higher or lower than the extreme predicted scores.